helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/algorithm/ConstraintBasedAlgorithm.java

```diff
@@ -0,0 +1,109 @@
+    private final List<SoftConstraint> _softConstraints;
+
+    public ConstraintBasedAlgorithm(List<HardConstraint> hardConstraints, List<SoftConstraint> softConstraints) {
+        _hardConstraints = hardConstraints;
+        _softConstraints = softConstraints;
+    }
+
+    @Override
+    public OptimalAssignment calculate(ClusterModel clusterModel) {
+        OptimalAssignment currentAssignment = new OptimalAssignment();
+        Map<String, Set<AssignableReplica>> replicasByResource = clusterModel.getAssignableReplicaMap();
+        List<AssignableNode> nodes = new ArrayList<>(clusterModel.getAssignableNodes().values());
+        ClusterContext clusterContext = clusterModel.getContext();
+
+
+        for (String resource : replicasByResource.keySet()) {
+            for (AssignableReplica replica : replicasByResource.get(resource)) {
+                Optional<AssignableNode> bestNode = getNodeWithHighestPoints(replica, nodes, clusterContext, currentAssignment);
+                bestNode.ifPresent(node -> {
+                    currentAssignment.addAssignment(replica, node);
+                    // adjust the node's capacity and fault zone information
+                    node.assign(replica);
+                    clusterContext.addPartitionToFaultZone(node.getFaultZone(), resource, replica.getPartitionName());
+                });
+                // stop immediately if any replica cannot find best assignable node
+                if (currentAssignment.hasAnyFailure()) {
+                    LOG.info("The current assignment encountered failure, clean up and rolling back to origin cluster context");
+                    // roll-back previous assignments
+                    currentAssignment.getCurrentAssignment().entrySet()
+                            .forEach(replicaToNodes ->
+                                    replicaToNodes.getValue().forEach(node -> node.release(replicaToNodes.getKey()))
+                            );
+                    clusterContext.resetFaultZoneMap();
+                    return currentAssignment;
+                }
+            }
+        }
+
+        return currentAssignment;
+    }
+
+    private Optional<AssignableNode> getNodeWithHighestPoints(
+            AssignableReplica replica, List<AssignableNode> assignableNodes, ClusterContext clusterContext, OptimalAssignment currentAssignment) {
+        // Only record the first hard constraint failure reason of the node
+        Map<String, HardConstraint> failedHardConstraintOfNode = new HashMap<>();
+        List<AssignableNode> candidateNodes = assignableNodes.stream()
+                .filter(candidate -> _hardConstraints.stream()
+                        .allMatch(constraint -> {
+                            boolean isValid = constraint.isAssignmentValid(candidate, replica, clusterContext);
+                            if (!isValid) {
+                                failedHardConstraintOfNode.put(candidate.getInstanceName(), constraint);
```

Comment:
HardConstraint has a failure reason field, could you please use that as the error message?
And it could be better if we just put all the possible failure reasons here. According to my experience, it would be frustrated if one fixed the first failure reason then the second comes. So better to check everyone and include all the possible failure reasons.

Reply:
That's a good point. Whether we want to find out failure reasons or stop if one constraint fails, the 1st gives us more ability to debug but the second saves the performance when the end goal is to come up with an assignment. That's up to the design. In the rebalance environment world, in which cases do we need to manually fix the fail reasons of hard constraints? If it won't even happen, I would suggest failing early.

Reply:
I can think of an example, if both cpu and memory are full, the users probably want to know both issues in one call.

Let's ping our potential users about this case.